### PR TITLE
CRD OpenAPI publishing 1.13 backport

### PR DESCRIPTION
**Pull request by sttts**
Pre-requisite: https://github.com/openshift/cluster-config-operator/pull/33

### CODE REVIEW

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 1]
```diff
@@ -50,8 +50,17 @@ func ValidateCustomResource(customResource interface{}, validator *validate.Sche
 	return nil
 }
 
-// ConvertJSONSchemaProps converts the schema from apiextensions.JSONSchemaPropos to go-openapi/spec.Schema
+// ConvertJSONSchemaProps converts the schema from apiextensions.JSONSchemaPropos to go-openapi/spec.Schema.
 func ConvertJSONSchemaProps(in *apiextensions.JSONSchemaProps, out *spec.Schema) error {
+	return ConvertJSONSchemaPropsWithPostProcess(in, out, nil)
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: This change is coming from https://github.com/kubernetes/kubernetes/pull/74778

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import (
+	"strings"
+
+	"github.com/go-openapi/spec"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+)
+
+// ConvertJSONSchemaPropsToOpenAPIv2Schema converts our internal OpenAPI v3 schema
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go]: Ditto https://github.com/kubernetes/kubernetes/pull/74778

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion_test.go [hunk 3]
```diff
@@ -0,0 +1,841 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion_test.go]: Ditto https://github.com/kubernetes/kubernetes/pull/74778

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go [hunk 4]
```diff
@@ -133,7 +133,8 @@ func gvkConvert(gvk schema.GroupVersionKind) v1.GroupVersionKind {
 	}
 }
 
-func friendlyName(name string) string {
+// FriendlyName returns REST resource friendly naming
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go]: Why this change? I don't see this being used (I might missed it), upstream does not have it.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 5]
```diff
@@ -124,6 +126,14 @@ type GenericAPIServer struct {
 	swaggerConfig *swagger.Config
 	openAPIConfig *openapicommon.Config
 
+	// OpenAPIVersionedService controls the /openapi/v2 endpoint, and can be used to update the served spec.
+	// It is set during PrepareRun.
+	OpenAPIVersionedService *handler.OpenAPIService
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: This is coming from https://github.com/kubernetes/kubernetes/pull/74098/

Comment (soltysh) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: I'm missing some parts of that PR ^, especially tests.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go [hunk 6]
```diff
@@ -18,9 +18,11 @@ package routes
 
 import (
 	restful "github.com/emicklei/go-restful"
+	"github.com/go-openapi/spec"
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go]: Ditto https://github.com/kubernetes/kubernetes/pull/74098

File: vendor/k8s.io/kubernetes/test/utils/crd/crd_util.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
+/*
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/test/utils/crd/crd_util.go]: These utils are coming from some bigger refactor, but I guess we can just keep them together with this particular commit.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go (hunk 6 above)
Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go]: fixed

**Comment by sttts** [comment]
https://github.com/openshift/cluster-config-operator/pull/33 merged.

/retest

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go [hunk 8]
```diff
@@ -38,6 +40,16 @@ func (oa OpenAPI) Install(c *restful.Container, mux *mux.PathRecorderMux) (*hand
 	if err != nil {
 		klog.Fatalf("Failed to register open api spec for root: %v", err)
 	}
+
+	// we shadow ClustResourceQuotas with a CRD. This loop removes all CRQ paths
+	// from the OpenAPI spec such that they don't conflict with the CRD
+	// apiextensions-apiserver spec during merging.
+	for pth := range spec.Paths.Paths {
+		if strings.HasPrefix(pth, "/apis/quota.openshift.io/v1/clusterresourcequotas") {
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go]: why not expose the list from our other carry patch here? so people don't have to maintain two lists

**Comment by sttts** [comment]
/retest

/hold cancel
as discussed in Slack: we will make the list global in a follow-up.

**Comment by mfojtik** [comment]
/retest

OLM flake

**Comment by deads2k** [comment]
this is needed for beta to be able to see the resources that we expose.  merging.

**Pull request merged**